backend/service/src/universal_codes/mod.rs

```diff
@@ -107,6 +122,32 @@ impl UniversalCodesService {
         }
     }
 
+    pub async fn product_by_code(
+        &self,
+        code: String,
+    ) -> Result<Option<Entity>, UniversalCodesServiceError> {
+        let result = dgraph::entity_with_parents_by_code(&self.client, code.clone())
+            .await
+            .map_err(|e| UniversalCodesServiceError::InternalError(e.message().to_string()))?;
+
+        let entity_with_parents = match result {
+            Some(entity) => entity,
+            None => return Ok(None),
+        };
+
+        if entity_with_parents.r#type == "Product" {
+            return Ok(Some(entity_with_parents));
+        }
+        // search through parents to find the parent code that matches the code
+
+        let product_code = match extract_product_code_from_entity(&entity_with_parents) {
+            Some(code) => code,
+            None => return Ok(None),
+        };
+
+        self.entity_by_code(product_code).await
```

Comment:
Could sort of re-create the heirachy for this particular code without another query, however we might miss other branches of the tree, so we need to do a new query here.